pkg/apis/openebs.io/upgrade/v1alpha2/types.go

```diff
@@ -0,0 +1,195 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=upgradetask
+// +k8s:openapi-gen=true
+
+// UpgradeTask represents an upgrade task
+type UpgradeTask struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec i.e. specifications of the UpgradeTask
+	Spec UpgradeTaskSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+	// Status of UpgradeTask
+	Status UpgradeTaskStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// UpgradeTaskSpec is the properties of an upgrade task
+type UpgradeTaskSpec struct {
+	// FromVersion is the current version of the resource.
+	FromVersion string `json:"fromVersion" protobuf:"bytes,1,name=fromVersion"`
+	// ToVersion is the upgraded version of the resource. It should be same
+	// as the version of control plane components version.
+	ToVersion string `json:"toVersion" protobuf:"bytes,2,name=toVersion"`
+	// Flags contains the optional flags that can be passed during upgrade.
+	Flags Flags `json:"flags,omitempty" protobuf:"bytes,3,name=flags"`
+	// ResourceSpec contains the details of the resource that has to upgraded.
+	ResourceSpec `json:",inline" protobuf:"bytes,4,name=resourceType"`
+	// ImagePrefix contains the url prefix of the image url. This field is
+	// optional. If not present upgrade takes the previously present ImagePrefix.
+	ImagePrefix string `json:"imagePrefix" protobuf:"bytes,5,name=imagePrefix"`
+	// ImageTag contains the customized tag for ToVersion if any. This field is
+	// optional. If not present upgrade takes the ToVersion as the ImageTag
+	ImageTag string `json:"imageTag" protobuf:"bytes,6,name=imageTag"`
+}
+
+// Flags provides additional optional arguments
+type Flags struct {
+	Timeout int `json:"timeout,omitempty" protobuf:"varint,1,opt,name=resourceType"`
+}
+
+// ResourceSpec is the type of resource which is to be upgraded.
+// Exactly one of its members must be set.
+type ResourceSpec struct {
+	// JivaVolume contains the details of the jiva volume to be upgraded
+	JivaVolume *JivaVolume `json:"jivaVolume,omitempty" protobuf:"bytes,1,opt,name=jivaVolume"`
+	// CStorVolume contains the details of the cstor volume to be upgraded
+	CStorVolume *CStorVolume `json:"cstorVolume,omitempty" protobuf:"bytes,2,opt,name=cstorVolume"`
+	// CStorPool contains the details of the cstor pool to be upgraded
+	CStorPool *CStorPool `json:"cstorPool,omitempty" protobuf:"bytes,3,opt,name=cstorPool"`
+	// StoragePoolClaim contains the details of the storage pool claim to be upgraded
+	StoragePoolClaim *StoragePoolClaim `json:"storagePoolClaim,omitempty" protobuf:"bytes,3,opt,name=storagePoolClaim"`
+}
```

Comment:
Does above spec imply all the entites must be specified? Or does it imply only one of the entities is required?

Reply:
only one should be specified